skynet/dgpu/daemon.py

```diff
@@ -32,274 +29,182 @@ def convert_reward_to_int(reward_str):
-
-                    except BaseException as err:
-                        logging.exception('Failed to serve model request !?\n')
-                        # traceback.print_exc()  # TODO? <- replaced by above ya?
-                        await self.conn.cancel_work(rid, str(err))
-
-                    finally:
-                        return True
-
-        # TODO, i would inverse this case logic to avoid an indent
-        # level in above block ;)
-        else:
-            logging.info(f'request {rid} already beign worked on, skip...')
-
-    # TODO, as per above on `.maybe_serve_one()`, it's likely a bit
-    # more *trionic* to define this all as a module level task-func
-    # which operates on a `daemon: SkynetDGPUDaemon`?
-    #
-    # -[ ] keeps tasks-as-funcs style prominent
-    # -[ ] avoids so much indentation due to methods
-    async def serve_forever(self):
+        logging.warning('model not blacklisted!, skip...')
+        return
+
+    results = [res['request_id'] for res in conn._tables['results']]
+
+    # if worker already produced a result for this request
+    if rid in results:
+        logging.info(f'worker already submitted a result for request #{rid}, skip...')
+        return
+
+    statuses = conn._tables['requests'][rid]
+
+    # skip if workers in non_compete already on it
+    competitors = set((status['worker'] for status in statuses))
+    if bool(config.non_compete & competitors):
+        logging.info('worker in configured non_compete list already working on request, skip...')
+        return
+
+    # resolve the ipfs hashes into the actual data behind them
+    inputs = []
+    raw_inputs = req['binary_data'].split(',')
+    if raw_inputs:
+        logging.info(f'fetching IPFS inputs: {raw_inputs}')
+
+    retry = 3
+    for _input in req['binary_data'].split(','):
+        if _input:
+            for r in range(retry):
+                try:
+                    # user `GPUConnector` to IO with
+                    # storage layer to seed the compute
+                    # task.
+                    img = await conn.get_input_data(_input)
+                    inputs.append(img)
+                    logging.info(f'retrieved {_input}!')
+                    break
+
+                except BaseException:
+                    logging.exception(
+                        f'IPFS fetch input error !?! retries left {retry - r - 1}\n'
+                    )
+
+    # compute unique request hash used on submit
+    hash_str = (
+        str(req['nonce'])
+        +
+        req['body']
+        +
+        req['binary_data']
+    )
+    logging.debug(f'hashing: {hash_str}')
+    request_hash = sha256(hash_str.encode('utf-8')).hexdigest()
+    logging.info(f'calculated request hash: {request_hash}')
+
+    params = body['params']
+    total_step = params['step'] if 'step' in params else 1
+    model = body['params']['model']
+    mode = body['method']
+
```

Comment:
why are we using mode in one place, and method in another?

Reply:
Again, compatibility with old request format, trying to converge into "mode"  in the future